primitives/tracing/src/proxy.rs

```diff
@@ -0,0 +1,113 @@
+use std::cell::RefCell;
+use rental;
+
+use crate::span_dispatch;
+
+const MAX_SPANS_LEN: usize = 1000;
+
+thread_local! {
+	static PROXY: RefCell<TracingProxy> = RefCell::new(TracingProxy::new());
+}
+
+pub fn create_registered_span(target: &str, name: &str) -> u64 {
+	PROXY.with(|proxy| proxy.borrow_mut().create_registered_span(target, name))
+}
+
+pub fn exit_span(id: u64) {
+	PROXY.with(|proxy| proxy.borrow_mut().exit_span(id));
+}
+
+rental! {
+	pub mod rent_span {
+		#[rental]
+		pub struct SpanAndGuard {
+			span: Box<tracing::Span>,
+			guard: tracing::span::Entered<'span>,
+		}
+	}
+}
+
+/// Requires a tracing::Subscriber to process span traces,
+/// this is available when running with client (and relevant cli params)
+pub struct TracingProxy {
+	next_id: u64,
+	spans: Vec<(u64, rent_span::SpanAndGuard)>,
+}
+
+impl Drop for TracingProxy {
+	fn drop(&mut self) {
+		while let Some((_, mut sg)) = self.spans.pop() {
+			sg.rent_all_mut(|s| { s.span.record("is_valid_trace", &false); });
+		}
+	}
+}
+
+impl TracingProxy {
+	pub fn new() -> TracingProxy {
+		let spans: Vec<(u64, rent_span::SpanAndGuard)> = Vec::new();
+		TracingProxy {
+			// Span ids start from 1 - we will use 0 as special case for unregistered spans
+			next_id: 1,
```

Comment:
Why not use `Option<u64>` to emphasize that there is a special value that is handled differently?

Reply:
This should be possible - but do you know if there would be any overhead in wasm if using `Option<u64>` vs `u64`?

Reply:
not an expert here, but even if there is I don't think it's worth considering. Like I'm not against using `0` as special value, but it feels cplusplusy to me, we have nice constructs in Rust that are better to express something like this:)